1188 External event triggers

hjoliver opened the pull request:
Prototype implementation of #1188.

@matthewrmshin et. al. - Not for final review yet, please consider the implementation in light of my explanatory comments in https://github.com/cylc/cylc/issues/1188#issuecomment-73492947. Note in particular the working example suite in ```examples/satellite/ext-triggers```.  I'll refrain from documenting this or tidying it up until we agree that this is the right approach.

[update: close #1188]

matthewrmshin commented:
@arjclark reassign to you to comment, as I'll be unavailable for some time in the next few weeks.

hjoliver commented:
(ref: email to @kaday and @matthewrmshin Feb 17) - Note the location of external trigger config under ```[runtime][[TASK]]```.  To maintain a strict separation between "scheduling" (affecting *when* a task runs) and "runtime" (affecting *what* runs exactly), these should be configured under ```[scheduling]```.  However, it's not clear to me how to do that cleanly without using ```[scheduling][[TASK]]``` sections, as external trigger messages are not concise like other special task config (e.g. "sequential" is just a list of task names, and "clock-triggered" is task names plus a short offset string).  More generally I wonder if we should relax this constraint and move even clock-triggers to task runtime sections, as that would undoubtedly be the easiest place to put them.  Thoughts?

arjclark commented:
Ok, will take a look. @hjoliver - I've emailed you re. turnaround.

arjclark commented on bin/cylc-external-trigger:
```diff
@@ -0,0 +1,58 @@
+
+event_id = None
+
+if len(args) < 1 or len(args) > 3:
```
Think I prefer `len(args) == 0` rather than ` < 1 `

arjclark commented on lib/cylc/config.py:
```diff
@@ -1380,12 +1391,15 @@ def generate_taskdefs( self, line, left_nodes, right, section, seq,
-                        self.taskdefs[name].outputs.append(outp)
-
+                # register explicit output messages
+                for lbl,msg in self.cfg['runtime'][name]['outputs'].items():
```
I guess the if isn't being used any more as iterating over nothing will have the same end result?

hjoliver replied:
Yes. I think the if test was required back in the ConfigObj days, but now the sections exist in the nested config dict even if they contain no items.

arjclark commented on lib/cylc/external_trigger.py:
```diff
@@ -0,0 +1,181 @@
+class PyroClient(object):
+    """Network client for sending external triggers to a suite."""
+
+    # TODO - configurable network retry parameters?
```
Sounds like a sensible idea - depending on the connection etc for these things longer timeouts/variable try numbers may be appropriate.

arjclark replied:
Override-able, maybe as arguments to the CLI command?

hjoliver replied:
> Override-able, maybe as arguments to the CLI command?

Good idea, I'll do that.

arjclark commented on lib/cylc/external_trigger.py:
```diff
@@ -0,0 +1,181 @@
+                self.pyro_proxy.register(self.event_msg, self.event_id)
+            except Pyro.errors.NamingError as exc:
+                print >> sys.stderr, exc
+                print "Send message: try %s of %s failed: %s" % (
```
This message could be a class level variable as it's used several times e.g. `ERROR_FAILED_SEND="Send message: try %s of %s failed: %s"`.

arjclark commented on lib/cylc/external_trigger.py:
```diff
@@ -0,0 +1,181 @@
+        # Note this has to allow multiple same-message triggers to be queued
+        # and only used one at a time.
+
+        # TODO - show in "show pre-requisites and outputs"
```
Is this still to do?

arjclark commented on lib/cylc/task_pool.py:
```diff
@@ -669,6 +675,10 @@ def any_task_failed(self):
 
+    def match_external_triggers(self):
+        for itask in self.get_tasks(incl_runahead=False):
+            self.external_trigger_broker.retrieve(itask)
```
maybe an if test along lines of
```
if itask.external_triggers:
```

arjclark commented:
Some preliminary code comments made. I'm not sure I like the location of these triggers under the runtime section as it hides them away from the other graphing parts (which ultimately they are part of)

Could we not do something like:

```
[scheduling]
    [[special tasks]]
        external-triggered = """foo(a, 'new dataset ready for processing'), \
                                            foo(b, 'and it is good') 
```

arjclark commented:
Am getting failures on this branch for ./tests/cylc-get-config/00-simple.t:

```
@@ -14,7 +14,6 @@
    pre-command scripting = 
    post-command scripting = 
    inherit = VAR, PARALLEL
-   [[[external triggers]]]
    [[[event hooks]]]
       submission timeout handler = 
       submitted handler =
```

hjoliver commented:
>  I'm not sure I like the location of these triggers under the runtime section ...

This relates to my comment above https://github.com/cylc/cylc/pull/1316#issuecomment-74580291 and to #1364. I'll just go with your suggestion in the first instance, for this pull request.

(hoping to get back to this tomorrow).

hjoliver commented:
@arjclark - I think I've addressed all of your points.  If you agree now, I just need to add some documentation, a test, and perhaps style-fixes to other files touched by this.

hjoliver commented:
(big PEP8 fix, sorry :grimacing:)

hjoliver commented:
(note, I've rebased this branch and fixed one problem; I'll push the changes up tomorrow)

hjoliver commented:
(just documentation to do now)

hjoliver commented:
Rebased onto master.  TODO:
 * a couple of tests are failing for trivial reasons,
 * documentation.

hjoliver commented:
(branch updated, test battery passes; still TODO - document external triggers).

hjoliver commented:
@arjclark - documentation done.  Sorry it took so long.  If you've forgotten the details I suggest a quick read of the new documentation section, and a look at the example suite.  Since your review above the only significant changes are the documentation, a rebase to master, and one fix for compatibility with another merge (f474577).   This is good to go now IMO (and we want to use it at NIWA).

hjoliver commented:
(this branch now requires a non-trivial update due to conflicts with the network code refactoring of #1355 :grimacing: )

arjclark commented:
Oh dear, will wait for the update. Am now mostly returned to work (just moved into a new house so got workmen coming and going).

hjoliver commented:
@arjclark - rebased onto master, and adapted to recent changes there.  Test battery passes in my environment.

arjclark commented on doc/cug.tex:
```diff
@@ -3789,30 +3789,124 @@ \subsubsection{Trigger Types}
+because it is beyond the suite stop point. Consequently cylc prevents tasks
+from spawning successors that depend on other tasks beyond the stop point.
+
+\paragraph{Clock  Triggers}
```
Extra space

arjclark commented on doc/cug.tex:
```diff
@@ -3789,30 +3789,124 @@ \subsubsection{Trigger Types}
+Note that cylc does not directly support ``filesystem events'' because systems
+such as \lstinline=inotify= on Linux are not portable. However, external
+watcher processes can of course use inotify to detect filesystem events, if
+that is appropriate, before triggering the suite.
```
perhaps "triggering tasks in a suite" rather than "the suite"

arjclark commented on doc/cug.tex:
```diff
@@ -3789,30 +3789,124 @@ \subsubsection{Trigger Types}
+Note that only one task in a suite can trigger off a given external message.
+This is no restriction however because other tasks can trigger off the
+externally triggered task.
+
```
Presumably can also only issue external triggers to suites for which you have the passphrase? If so, worth mentioning as people might think to use this mechanism to trigger off other people's systems/suites etc.

arjclark commented on examples/satellite/task-polling/suite.rc:
```diff
@@ -1,10 +1,12 @@
-Each successive integer cycle retrieves and processes the next
-arbitrarily timed and arbitrarily labelled dataset, in parallel
-with previous cycles if the data comes in quickly."""
+Successive cycle points retrieve and processe the next arbitrarily timed and
```
extra "e" on end of "process"

arjclark commented on examples/satellite/task-polling/suite.rc:
```diff
@@ -1,10 +1,12 @@
-arbitrarily timed and arbitrarily labelled dataset, in parallel
-with previous cycles if the data comes in quickly."""
+Successive cycle points retrieve and processe the next arbitrarily timed and
+labelled dataset, in parallel if the data comes in quickly. This variant of the
```
either extraneous comma after dataset or a missing one after parallel

hjoliver replied:
No, the meaning is: datasets will be processed in parallel if they come in quickly enough, otherwise not.

arjclark replied:
Ah, ok, not the way I read it :)

arjclark commented on lib/cylc/network/ext_trigger.py:
```diff
@@ -0,0 +1,173 @@
+        return sent
+
+
+        #try:
```
Can we remove the commented out code?

hjoliver replied:
Oops, yes.

arjclark commented:
@hjoliver - looks ok to me. Test battery passing in my environment. A few minor tweaks requested above. Otherwise good to go unless @matthewrmshin has anything to add?

hjoliver commented:
@arjclark - thanks, review points addressed.  Only one minor point of dissension (the supposedly extraneous comma).

arjclark commented:
Looks good to me. @matthewrmshin has told me he has nothing to add. Test battery passing in my environment.
